File: RaceLaneManager/Model/Race.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using System;

namespace RaceLaneManager.Model
{
    public enum RaceState
    {
        NotStarted,
        Racing,
        Done
    }

    public interface IRace
    {
        int RaceNumber { get; }
        RaceState State { get; }
        IEnumerable<ILaneAssignment> LaneAssignments { get; }
    }

    public class Race : IRace
    {
        [JsonIgnore]
        public List<LaneAssignment> LaneAssignmentData { get; set; }
        public int RaceNumber { get; set; }
        public RaceState State { get; set; }
        public IEnumerable<ILaneAssignment> LaneAssignments { get { return this.LaneAssignmentData; } }

        public Race()
        {
            this.LaneAssignmentData = new List<LaneAssignment>();
        }
    }
}

File: website/models/race.ts
enum RaceState {
    NotStarted,
    Racing,
    Done
}

class Race {
    raceNum: number;
    state: RaceState;
    laneAssignments: Array<LaneAssignment>;

    static fromPayload(payload: any): Race {
        return {
            raceNum: payload.RaceNumber,
            state: payload.State,
            laneAssignments: LaneAssignment.getLaneAssignmentsFromPayload(payload.LaneAssignments)
        };
    }

    static toPayload(race: Race): any {
        return {
            'RaceNumber': race.raceNum,
            'State': race.state,
            'LaneAssignments': race.laneAssignments
        };
    }
    
    static getRacesFromPayload(payload: any): Array<Race> {
        let races: Array<Race> = new Array<Race>();
        
        for (let r of payload) {
            races.push(Race.fromPayload(r));
        }

        return races;
    }
}

